### PR TITLE
autosuspend: init at 3.1.3

### LINKED ISSUE #132392: Packaging request: Autosuspend
**Project description**
`autosuspend` is a python daemon that suspends a system if certain conditions are met, or not met. This enables a server to sleep in case of inactivity without depending on the X infrastructure usually used by normal desktop environments.

**Metadata**

* homepage URL: https://autosuspend.readthedocs.io/en/latest/
* source URL: https://github.com/languitar/autosuspend
* license: gpl2
* platforms: linux

### PR DESCRIPTION
**Pull request by maggicl**
###### Motivation for this change
The [autosuspend](https://github.com/NixOS/nixpkgs/issues/132392) package is missing from nixpkgs and this PR includes an appropriate derivation and module to use it.
Resolves #132392 

###### Things done
Writing a basic derivation and module definition for autosuspend.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS (package not compatible)
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking (not a package update)
  - [ ] (Module updates) Added a release notes entry if the change is significant (not a module update)
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md) (misses valid maintainer).

### CODE REVIEW

**Comment by maggicl** [comment]
First contribution to Nix/Nixos, hope everything follows the conventions and it is at least decent quality. Couple of notes:

- building the man pages requires two pip packages that are not in nixpkgs, and they are currently built in a let expression surrounding the derivation. The proper way to do it is probably to package these two dependencies properly, but I'm unsure if I should open other PRs for them or do it here.
- I'm not sure what to put in the "maintainer" field, i.e. if I should add myself to one of the maintainer groups or someone should "adopt" the package.

Feedback is appreciated :+1:

File: pkgs/tools/misc/autosuspend/default.nix [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  };
+in
+  with python3.pkgs; buildPythonApplication rec {
+    pname = "autosuspend";
+    version = "3.1.3";
+
+    src = fetchFromGitHub {
+      owner = "languitar";
+      repo = "autosuspend";
+      rev = "6adc8f812ebcaeae83d2b0725ca37d7b433b7f4a";
```
Comment (aanderse) [inline pkgs/tools/misc/autosuspend/default.nix:43]:
```suggestion
      rev = "v${version}";
```

File: nixos/modules/services/misc/autosuspend.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.autosuspend;
+  settingsFormat = pkgs.formats.ini {};
+in {
+  options.services.autosuspend = {
+    enable = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/misc/autosuspend.nix:10]: Are you familiar with `mkEnableOption` at all?

**Review by aanderse — COMMENTED** [review]
Hi there! Welcome, and thanks for choosing to contribute to the project.

> First contribution to Nix/Nixos

Woo! 🚀 

> The proper way to do it is probably to package these two dependencies properly, but I'm unsure if I should open other PRs for them or do it here.

> Here seems reasonable.

If you're willing to become a maintainer that is awesome. You can add yourself to the maintainer list. Please do that in a separate commit and have that commit come first.

I don't generally review packages, just modules, but don't worry... someone will be along shortly to review the packages I'm sure.

I have left a few comments for discussion. Don't hesitate to ask for any advice/help/etc...

File: pkgs/tools/misc/autosuspend/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    disabled = pythonOlder "3.6";
+
+    buildInputs = [ setuptools-scm setuptools ];
+    propagatedBuildInputs = [ sphinx ];
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "1wq9yp66xq9scazzc05vgh1n84cyf0gl64vgf2fc1i3sdirr8ll4";
+    };
+  };
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:34]: Please move the two packages into pythonPackages.

File: pkgs/tools/misc/autosuspend/default.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+, fetchFromGitHub
+}:
+
+let
+  sphinx-autodoc-typehints = with python3.pkgs; buildPythonPackage rec {
+    pname = "sphinx-autodoc-typehints";
+    version = "1.12.0";
+    disabled = pythonOlder "3.6";
+
+    buildInputs = [ setuptools-scm setuptools ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:14]:
```suggestion
    nativeBuildInputs = [ setuptools-scm ];
    buildInputs = [ setuptools ];
```

File: pkgs/tools/misc/autosuspend/default.nix [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      inherit pname version;
+      sha256 = "0llzia8clzq7vvpwf0ifr1cxjk73fjg3d79rn60p5170vgcifdhr";
+    };
+  };
+  sphinx-issues = with python3.pkgs; buildPythonPackage rec {
+    pname = "sphinx-issues";
+    version = "1.2.0";
+    disabled = pythonOlder "3.6";
+
+    buildInputs = [ setuptools-scm setuptools ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:27]:
```suggestion
    nativeBuildInputs = [ setuptools-scm ];
    buildInputs = [ setuptools ];
```

File: pkgs/tools/misc/autosuspend/default.nix [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+      dateutil
+      tzlocal
+      freezegun
+      jsonpath-ng
+      portalocker
+      psutil
+    ];
+
+    buildInputs = [
+      pytest
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:62]:
```suggestion
      pytestCheckHook
```

File: pkgs/tools/misc/autosuspend/default.nix [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      cp doc/build/man/autosuspend.1 $out/share/man/man1
+      cp doc/build/man/autosuspend.conf.5 $out/share/man/man5
+    '';
+
+    meta = with lib; {
+      homepage = "https://autosuspend.readthedocs.io";
+      description = "Python daemon that suspends a system if certain conditions are met, or not met";
+      license = licenses.gpl2;
+      platforms = platforms.linux;
+      # maintainers = with maintainers; [ ];
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:85]: You need to add this.

File: pkgs/tools/misc/autosuspend/default.nix [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      sphinx-autodoc-typehints
+      sphinxcontrib_plantuml
+      recommonmark
+    ];
+
+    postInstall = ''
+      python3 setup.py build_sphinx -a -b man
+      mkdir -p $out/share/man/{man1,man5}
+      cp doc/build/man/autosuspend.1 $out/share/man/man1
+      cp doc/build/man/autosuspend.conf.5 $out/share/man/man5
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:77]: Use installManPage.

File: pkgs/tools/misc/autosuspend/default.nix [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      python-dbusmock
+      sphinx
+      sphinx-issues
+      sphinx-autodoc-typehints
+      sphinxcontrib_plantuml
+      recommonmark
+    ];
+
+    postInstall = ''
+      python3 setup.py build_sphinx -a -b man
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:74]:
```suggestion
      ${python3.interpreter} setup.py build_sphinx -a -b man
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/autosuspend/default.nix [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    buildInputs = [
+      pytest
+      pytest-mock
+      pytest-httpserver
+      python-dbusmock
+      sphinx
+      sphinx-issues
+      sphinx-autodoc-typehints
+      sphinxcontrib_plantuml
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:69]:
```suggestion
      sphinxcontrib-plantuml
```

File: pkgs/tools/misc/autosuspend/default.nix (hunk 6 above)
Comment (Artturin) [reply pkgs/tools/misc/autosuspend/default.nix:62]: also, the pytest related package can probably go into checkInputs

File: nixos/modules/services/misc/autosuspend.nix [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  };
+
+  config = mkIf cfg.enable {
+    environment.etc."autosuspend.conf".source =
+      settingsFormat.generate "autosuspend-conf.ini" cfg.settings;
+
+    environment.etc."autosuspend-logging.conf".source =
+      settingsFormat.generate "autosuspend-logging-conf.ini" cfg.loggingSettings;
+
+    systemd.services.autosuspend = {
```
[the remark this answers is not in the document]
Comment (maggicl) [reply nixos/modules/services/misc/autosuspend.nix:46]: Absolutely none. I thought using system.services would have been more appropriate than using the original file. I wouldn't know how to simply import the upstream unit file, though I'm definitely a newbie at this.

File: nixos/modules/services/misc/autosuspend.nix (hunk 2 above)
Comment (maggicl) [reply nixos/modules/services/misc/autosuspend.nix:10]: I wasn't. To be honest I just followed an example from Nix Pills on how to create a basic module, and mkEnableOption wasn't covered there

File: nixos/modules/services/misc/autosuspend.nix (hunk 11 above)
[the remark this answers is not in the document]
Comment (Artturin) [reply nixos/modules/services/misc/autosuspend.nix:40]:
```
systemd = { packages = [ pkgs.autosuspend ];
# https://github.com/NixOS/nixpkgs/issues/81138
services.autosuspend.wantedBy = [ "multi-user.target" ];
};
```

File: pkgs/tools/misc/autosuspend/default.nix (hunk 10 above)
Comment (maggicl) [reply pkgs/tools/misc/autosuspend/default.nix]: I had to edit the entry in `pkgs/top-level/python-packages.nix` and change the underscore in a hyphen there in order to make it work. Running a quick `git grep` shows that no other package in nixpkgs depends on `sphinxcontrib-plantuml`, but still I renamed a package and I would like to make sure it's alright

File: nixos/modules/services/misc/autosuspend.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/misc/autosuspend.nix]: It's just a convenience function, not required.

File: nixos/modules/services/misc/autosuspend.nix [hunk 12]
```diff
@@ -0,0 +1,46 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.autosuspend;
+  settingsFormat = pkgs.formats.ini {};
+in {
+  options.services.autosuspend = {
+    enable = mkEnableOption "If enabled, NixOS will start the autosuspend service.";
```
Comment (aanderse) [inline nixos/modules/services/misc/autosuspend.nix:10]:
```suggestion
    enable = mkEnableOption "the autosuspend service";
```

This function adds a period to the end and sets up a nice prefix.

**Review by jonringer — COMMENTED** [review]
To comply with [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes) please have the commit message name be of the format
```
<pkg-name>: <subject-line>
```
for more examples, please look at https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes

in your case, you should have a commit per package and module added.

File: pkgs/tools/misc/autosuspend/default.nix (hunk 10 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/autosuspend/default.nix]: It was added with #132421

File: maintainers/maintainer-list.nix [hunk 13]
```diff
@@ -12234,4 +12234,10 @@
     github = "zupo";
     githubId = 311580;
   };
+  maggicl = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:12237]: Please sort this alphabetically.

File: pkgs/tools/misc/autosuspend/default.nix [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    sphinx
+    sphinx-issues
+    sphinx-autodoc-typehints
+    sphinxcontrib-plantuml
+    recommonmark
+  ];
+
+  checkInputs = [
+    pytestCheckHook
+    pytest-cov
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/autosuspend/default.nix:46]:
```suggestion
```
please patch out the usage of cov.

**Comment by WolfangAukang** [comment]
Unfortunately, this [package](https://github.com/NixOS/nixpkgs/tree/master/pkgs/os-specific/linux/autosuspend) has been already merged into nixpkgs.

I see you created a module for it, if you want, we can try to merge your module.

**Comment by NickCao** [comment]
modules is in https://github.com/NixOS/nixpkgs/pull/210666

**Pull request closed without merging**